report/id.go

```diff
@@ -200,22 +200,25 @@ func ParseOverlayNodeID(id string) (overlayPrefix string, peerName string) {
 // ParseNodeID produces the host ID and remainder (typically an address) from
 // a node ID. Note that hostID may be blank.
 func ParseNodeID(nodeID string) (hostID string, remainder string, ok bool) {
-	fields := strings.SplitN(nodeID, ScopeDelim, 2)
-	if len(fields) != 2 {
+	first := strings.Index(nodeID, ScopeDelim)
+	if first == -1 {
 		return "", "", false
 	}
-	return fields[0], fields[1], true
+	return nodeID[:first], nodeID[first+1:], true
 }
 
 // ParseEndpointNodeID produces the scope, address, and port and remainder.
-// Note that hostID may be blank.
+// Note that scope may be blank.
 func ParseEndpointNodeID(endpointNodeID string) (scope, address, port string, ok bool) {
-	fields := strings.SplitN(endpointNodeID, ScopeDelim, 3)
-	if len(fields) != 3 {
+	first := strings.Index(endpointNodeID, ScopeDelim)
+	if first == -1 {
 		return "", "", "", false
 	}
-
-	return fields[0], fields[1], fields[2], true
+	second := strings.Index(endpointNodeID[first+1:], ScopeDelim)
+	if second == -1 {
+		return "", "", "", false
+	}
+	return endpointNodeID[:first], endpointNodeID[first+1 : first+1+second], endpointNodeID[first+1+second+1:], true
```

Comment:
Please add a comment explaining that we aren't using `strings.SplitN` in this function because it makes more allocations. (Also, is it just the allocation of the fields slice we are saving here, or is there more to it?)

Reply:
Yes, just the fields slice.

```
$ go test -tags 'unsafe netgo' -run=x -bench=TopologyList -benchtime=1ms -memprofile=mem.out -memprofilerate=111 -bench-report-file=prod-report1.json
$ go tool pprof -cum -alloc_objects -peek=NodeID mem.out 
...
                                             64437 64.45% |   github.com/weaveworks/scope/render.connectionJoin.Render
                                             35544 35.55% |   github.com/weaveworks/scope/render.pseudoNodeID
         0     0%     0%      99981  3.50%                | github.com/weaveworks/scope/report.ParseEndpointNodeID
                                             99981   100% |   strings.SplitN
----------------------------------------------------------+-------------
                                             17532 64.24% |   github.com/weaveworks/scope/report.ExtractHostID
                                              9758 35.76% |   github.com/weaveworks/scope/render.endpoints2Processes.Render
         0     0%     0%      27290  0.96%                | github.com/weaveworks/scope/report.ParseNodeID
                                             27290   100% |   strings.SplitN
...
$ go tool pprof -cum -alloc_objects -list=genSplit mem.out 
Total: 2855587
ROUTINE ======================== strings.genSplit in /usr/local/go/src/strings/strings.go
    128653     128653 (flat, cum)  4.51% of Total
         .          .    240:	if n < 0 {
         .          .    241:		n = Count(s, sep) + 1
         .          .    242:	}
         .          .    243:	c := sep[0]
         .          .    244:	start := 0
    128653     128653    245:	a := make([]string, n)
         .          .    246:	na := 0
         .          .    247:	for i := 0; i+len(sep) <= len(s) && na+1 < n; i++ {
         .          .    248:		if s[i] == c && (len(sep) == 1 || s[i:i+len(sep)] == sep) {
         .          .    249:			a[na] = s[start : i+sepSave]
         .          .    250:			na++
```